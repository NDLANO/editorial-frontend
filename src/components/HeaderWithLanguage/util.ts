/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { Node } from "@ndla/types-taxonomy";
import {
  toEditAudio,
  toEditConcept,
  toEditFrontPageArticle,
  toEditGloss,
  toEditImage,
  toEditLearningResource,
  toEditPodcast,
  toEditPodcastSeries,
  toEditTopicArticle,
} from "../../util/routeHelpers";

export const getTaxonomyPathsFromTaxonomy = (taxonomy?: Pick<Node, "paths">[], articleId?: number): string[] => {
  const taxPaths = taxonomy?.flatMap((t) => t.paths) ?? [];
  const articlePath = articleId ? `/article/${articleId}` : undefined;
  const paths = articlePath ? taxPaths.concat(articlePath) : taxPaths;
  return paths.filter((p): p is string => p !== undefined);
};

export const toMapping = {
  concept: toEditConcept,
  gloss: toEditGloss,
  audio: toEditAudio,
  "podcast-series": toEditPodcastSeries,
  podcast: toEditPodcast,
  image: toEditImage,
  "frontpage-article": toEditFrontPageArticle,
  standard: toEditLearningResource,
  "topic-article": toEditTopicArticle,
};

export type TranslatableType = keyof typeof toMapping;

export const translatableTypes: TranslatableType[] = [
  "audio",
  "concept",
  "gloss",
  "standard",
  "topic-article",
  "podcast",
  "image",
  "podcast-series",
  "frontpage-article",
];

export const createEditUrl = (id: number, locale: string, type: keyof typeof toMapping) => {
  return toMapping[type](id, locale);
};
