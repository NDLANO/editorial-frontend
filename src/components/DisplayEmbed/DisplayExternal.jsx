/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { injectT } from '@ndla/i18n';
import Types from 'slate-prop-types';
import './helpers/h5pResizer';
import handleError from '../../util/handleError';
import EditorErrorMessage from '../SlateEditor/EditorErrorMessage';
import DisplayExternalModal from './helpers/DisplayExternalModal';
import { fetchExternalOembed } from '../../util/apiHelpers';
import { EditorShape } from '../../shapes';
import { urlDomain, getIframeSrcFromHtmlString } from '../../util/htmlHelpers';
import { EXTERNAL_WHITELIST_PROVIDERS } from '../../constants';
import DeleteButton from '../DeleteButton';
import FigureButtons from '../SlateEditor/plugins/embed/FigureButtons';
import config from '../../config';
import { getH5pLocale } from '../H5PElement/h5pApi';

export class DisplayExternal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isEditMode: false,
    };
    this.onEditEmbed = this.onEditEmbed.bind(this);
    this.handleChangeVisualElement = this.handleChangeVisualElement.bind(this);
    this.openEditEmbed = this.openEditEmbed.bind(this);
    this.closeEditEmbed = this.closeEditEmbed.bind(this);
    this.getPropsFromEmbed = this.getPropsFromEmbed.bind(this);
  }

  componentDidMount() {
    this.getPropsFromEmbed();
  }

  componentDidUpdate(prevProps) {
    const { embed } = this.props;
    if (prevProps.embed.url !== embed.url || prevProps.embed.path !== embed.path) {
      this.getPropsFromEmbed();
    }
  }

  onEditEmbed(properties) {
    const { editor, node, embed } = this.props;

    if (properties.url !== embed.url || properties.path !== embed.path) {
      editor.setNodeByKey(node.key, {
        data: {
          ...properties,
        },
      });
      this.closeEditEmbed();
    }
  }

  async getPropsFromEmbed() {
    const { embed, language } = this.props;
    const domain = embed.url ? urlDomain(embed.url) : config.h5pApiUrl;
    const cssUrl = encodeURIComponent(`${config.ndlaFrontendDomain}/static/h5p-custom-css.css`);
    this.setState({ domain });

    if (embed.resource === 'external' || embed.resource === 'h5p') {
      try {
        let url = embed.url || `${domain}${embed.path}`;
        url = url.includes(config.h5pApiUrl)
          ? `${url}?locale=${getH5pLocale(language)}&cssUrl=${cssUrl}`
          : url;

        const data = await fetchExternalOembed(url);
        const src = getIframeSrcFromHtmlString(data.html);

        if (src) {
          this.setState({
            title: data.title,
            src,
            type: data.type,
            provider: data.providerName,
            height: data.height || '486px',
          });
        } else {
          this.setState({ error: true });
        }
      } catch (err) {
        handleError(err);
      }
    } else {
      this.setState({
        title: domain,
        src: embed.url,
        type: embed.resource,
        height: embed.height,
      });
    }
  }

  handleChangeVisualElement(providerName) {
    const { changeVisualElement } = this.props;
    if (changeVisualElement) {
      changeVisualElement(providerName);
    }
  }

  openEditEmbed(evt, providerName) {
    evt.preventDefault();
    this.handleChangeVisualElement(providerName);
    this.setState({ isEditMode: true });
  }

  closeEditEmbed(providerName) {
    this.handleChangeVisualElement(providerName);
    this.setState({ isEditMode: false });
  }

  render() {
    const { onRemoveClick, embed, t, language } = this.props;
    const { isEditMode, title, src, height, error, type, provider, domain } = this.state;

    const errorHolder = () => (
      <>
        <DeleteButton stripped onClick={onRemoveClick} />
        <EditorErrorMessage
          msg={
            error
              ? t('displayOembed.errorMessage')
              : t('displayOembed.notSupported', {
                  type,
                  provider,
                })
          }
        />
      </>
    );

    if (error) {
      return errorHolder();
    }

    // H5P does not provide its name
    const providerName = domain && domain.includes('h5p') ? 'H5P' : provider;

    const [allowedProvider] = EXTERNAL_WHITELIST_PROVIDERS.filter(whitelistProvider =>
      type === 'iframe'
        ? whitelistProvider.url.includes(domain)
        : whitelistProvider.name === providerName,
    );

    if (!allowedProvider) {
      return errorHolder();
    }

    if (!src || !type) {
      return <div></div>;
    }
    return (
      <div className="c-figure">
        <FigureButtons
          language={language}
          tooltip={t('form.external.remove', {
            type: providerName || t('form.external.title'),
          })}
          onRemoveClick={onRemoveClick}
          embed={embed}
          providerName={providerName}
          figureType="external"
          onEdit={
            allowedProvider.name
              ? evt => this.openEditEmbed(evt, allowedProvider.name.toLowerCase())
              : undefined
          }
        />
        <iframe
          ref={iframe => {
            this.iframe = iframe;
          }}
          src={src}
          height={allowedProvider.height || height}
          title={title}
          scrolling={type === 'iframe' ? 'no' : undefined}
          allowFullScreen={allowedProvider.fullscreen || true}
          frameBorder="0"
        />
        <DisplayExternalModal
          isEditMode={isEditMode}
          src={src}
          type={type}
          onEditEmbed={this.onEditEmbed}
          onClose={this.closeEditEmbed}
          allowedProvider={allowedProvider}
        />
      </div>
    );
  }
}

DisplayExternal.propTypes = {
  onRemoveClick: PropTypes.func,
  changeVisualElement: PropTypes.func,
  editor: EditorShape,
  node: Types.node,
  isIframe: PropTypes.bool,
  embed: PropTypes.shape({
    width: PropTypes.string,
    url: PropTypes.string,
    path: PropTypes.string,
    height: PropTypes.string,
    resource: PropTypes.string,
  }),
  language: PropTypes.string,
};

export default injectT(DisplayExternal);
