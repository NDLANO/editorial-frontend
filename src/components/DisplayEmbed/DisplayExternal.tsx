/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { useEffect, useState, useRef, HTMLAttributes } from "react";
import { useTranslation } from "react-i18next";
import { Transforms, Editor, Path } from "slate";
import { RenderElementProps } from "slate-react";
import "./helpers/h5pResizer";
import styled from "@emotion/styled";
import { IconButtonV2 } from "@ndla/button";
import { Pencil } from "@ndla/icons/action";
import { DeleteForever, Expandable } from "@ndla/icons/editor";
import DisplayExternalModal from "./helpers/DisplayExternalModal";
import SlateResourceBox from "./SlateResourceBox";
import config from "../../config";
import { EXTERNAL_WHITELIST_PROVIDERS } from "../../constants";
import { Embed, ExternalEmbed, H5pEmbed } from "../../interfaces";
import { fetchExternalOembed } from "../../util/apiHelpers";
import handleError from "../../util/handleError";
import { urlOrigin, getIframeSrcFromHtmlString, urlDomain } from "../../util/htmlHelpers";
import { getH5pLocale } from "../H5PElement/h5pApi";
import EditorErrorMessage from "../SlateEditor/EditorErrorMessage";
import { StyledDeleteEmbedButton, StyledFigureButtons } from "../SlateEditor/plugins/embed/FigureButtons";

const ApplyBoxshadow = styled.div`
  &[data-show-copy-outline="true"] {
    box-shadow: rgb(32, 88, 143) 0 0 0 2px;
  }
`;

const ExpandableButton = styled.div`
  position: absolute;
  bottom: 0px;
  right: 23px;
  cursor: pointer;
`;

type EmbedType = ExternalEmbed | H5pEmbed;

interface Props extends HTMLAttributes<HTMLDivElement> {
  pathToEmbed: Path;
  editor: Editor;
  embed: EmbedType;
  onRemoveClick: (event: React.MouseEvent) => void;
  language: string;
  active: boolean;
  isSelectedForCopy: boolean;
  attributes?: RenderElementProps["attributes"];
}

interface EmbedProperties {
  domain?: string;
  src?: string;
  title?: string;
  type?: string;
  height?: number | string;
  provider?: string;
}

const DisplayExternal = ({
  pathToEmbed,
  editor,
  embed,
  onRemoveClick,
  language,
  active,
  isSelectedForCopy,
  attributes,
  children,
}: Props) => {
  const [isEditMode, setIsEditMode] = useState(false);
  const [error, setError] = useState(false);
  const [properties, setProperties] = useState<EmbedProperties>({
    type: embed.resource,
  });
  const { t } = useTranslation();
  const prevEmbed = useRef<EmbedType>(embed);
  const [height, setHeight] = useState(0);
  const [isResizing, setIsResizing] = useState(false);
  const iframeWrapper = useRef(null);

  const getPropsFromEmbed = async () => {
    const origin = embed.url ? urlOrigin(embed.url) : config.h5pApiUrl;
    const domain = embed.url ? urlDomain(embed.url) : config.h5pApiUrl;
    const cssUrl = encodeURIComponent(`${config.ndlaFrontendDomain}/static/h5p-custom-css.css`);

    if (embed.resource === "external" || embed.resource === "h5p") {
      try {
        const base = embed.resource === "h5p" ? `${origin}${embed.path}` : embed.url;
        const url =
          config.h5pApiUrl && base.includes(config.h5pApiUrl)
            ? `${base}?locale=${getH5pLocale(language)}&cssUrl=${cssUrl}`
            : base;

        const data = await fetchExternalOembed(url);
        const src = getIframeSrcFromHtmlString(data.html);

        if (src) {
          setHeight(0);
          setProperties({
            ...properties,
            title: embed.title?.length ? embed.title : data.title,
            src,
            type: data.type,
            provider: data.providerName,
            height: data.height || "486px",
            domain: domain,
          });
        } else {
          setError(true);
        }
      } catch (err) {
        setError(true);
        handleError(err);
      }
    } else {
      // Update height if height of inserted element changes - otherwise reset height
      if (embed.height && prevEmbed.current.url === embed.url) {
        setHeight(Number(embed.height.replace(/\D/g, "")));
      } else {
        setHeight(0);
      }
      setProperties({
        ...properties,
        title: embed.title?.length ? embed.title : domain,
        src: embed.url,
        type: embed.resource,
        height: embed.height,
        domain: domain,
      });
    }
  };

  useEffect(() => {
    getPropsFromEmbed();
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    const prevEmbedElement: EmbedType = prevEmbed.current;

    if (prevEmbedElement.resource !== embed.resource) {
      getPropsFromEmbed();
    } else if (
      embed.resource === "h5p" &&
      prevEmbedElement.resource === "h5p" &&
      embed.path !== prevEmbedElement.path
    ) {
      getPropsFromEmbed();
    } else if (
      (embed.resource === "external" || embed.resource === "iframe") &&
      (prevEmbedElement.resource === "external" || prevEmbedElement.resource === "iframe") &&
      embed.url !== prevEmbedElement.url
    ) {
      getPropsFromEmbed();
    } else if (embed.title !== prevEmbedElement.title) {
      getPropsFromEmbed();
    }
    prevEmbed.current = embed;
  }, [embed]); // eslint-disable-line react-hooks/exhaustive-deps

  const openEditEmbed = (evt: React.MouseEvent) => {
    evt.preventDefault();
    setIsEditMode(true);
  };

  const closeEditEmbed = () => {
    setIsEditMode(false);
  };
  const onEditEmbed = (embedUpdates: Embed) => {
    Transforms.setNodes(editor, { data: { ...embedUpdates } }, { at: pathToEmbed });
    closeEditEmbed();
  };

  const showCopyOutline = isSelectedForCopy && (!isEditMode || !active);

  const errorHolder = () => (
    <EditorErrorMessage
      onRemoveClick={onRemoveClick}
      msg={
        error
          ? t("displayOembed.errorMessage")
          : t("displayOembed.notSupported", {
              type: properties.type,
              provider: properties.provider,
            })
      }
    />
  );

  if (error) {
    return errorHolder();
  }

  // H5P does not provide its name
  const providerName = properties.domain?.includes("h5p") ? "H5P" : properties.provider;

  const [allowedProvider] = EXTERNAL_WHITELIST_PROVIDERS.filter((whitelistProvider) =>
    properties.type === "iframe" && properties.domain
      ? whitelistProvider.url.includes(properties.domain)
      : whitelistProvider.name === providerName,
  );

  if (!allowedProvider) {
    return errorHolder();
  }

  if (!properties.src || !properties.type) {
    return <div />;
  }

  const handleResize = (mouseDownEvent: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    const startSize = height;
    const startPosition = mouseDownEvent.pageY;
    const minHeight = 100;

    setIsResizing(true);

    const onMouseMove = (mouseMoveEvent: MouseEvent) => {
      const elementHeight = startSize - startPosition + mouseMoveEvent.pageY;
      setHeight(elementHeight > minHeight ? elementHeight : minHeight);
    };

    const onMouseUp = (mouseUpEvent: MouseEvent) => {
      document.body.removeEventListener("mousemove", onMouseMove);
      const elementHeight = startSize - startPosition + mouseUpEvent.pageY;
      const updatedElementHeight = elementHeight > minHeight ? elementHeight : minHeight;

      const newData = {
        ...prevEmbed.current,
        height: `${updatedElementHeight}px`,
      };

      Transforms.setNodes(editor, { data: newData }, { at: pathToEmbed });
      setHeight(updatedElementHeight);
      setIsResizing(false);
    };

    document.body.addEventListener("mousemove", onMouseMove);
    document.body.addEventListener("mouseup", onMouseUp, { once: true });
  };

  return (
    <div {...attributes} className={"c-figure"} contentEditable={false}>
      <StyledFigureButtons data-white={true}>
        {allowedProvider.name && (
          <IconButtonV2
            aria-label={t("form.external.edit", {
              type: providerName || t("form.external.title"),
            })}
            title={t("form.external.edit", {
              type: providerName || t("form.external.title"),
            })}
            colorTheme="light"
            onClick={(evt) => {
              evt.preventDefault();
              evt.stopPropagation();
              openEditEmbed(evt);
            }}
          >
            <Pencil />
          </IconButtonV2>
        )}
        <StyledDeleteEmbedButton
          aria-label={t("form.external.remove", {
            type: providerName || t("form.external.title"),
          })}
          title={t("form.external.remove", {
            type: providerName || t("form.external.title"),
          })}
          colorTheme="danger"
          onClick={onRemoveClick}
          data-testid="remove-element"
        >
          <DeleteForever />
        </StyledDeleteEmbedButton>
      </StyledFigureButtons>
      {(embed.resource === "iframe" || embed.resource === "external") && embed.type === "fullscreen" ? (
        <ApplyBoxshadow data-show-copy-outline={showCopyOutline}>
          <SlateResourceBox embed={embed} language={language} />
        </ApplyBoxshadow>
      ) : (
        <ApplyBoxshadow
          ref={iframeWrapper}
          data-show-copy-outline={showCopyOutline}
          style={{ pointerEvents: isResizing ? "none" : "auto" }}
        >
          <iframe
            contentEditable={false}
            src={properties.src}
            height={height ? height : allowedProvider.height || properties.height}
            title={properties.title}
            scrolling={properties.type === "iframe" ? "no" : undefined}
            allowFullScreen={true}
            frameBorder="0"
          />
          {embed.resource === "iframe" && (
            <ExpandableButton role="button" onMouseDown={handleResize} aria-label={t("form.resize")}>
              <Expandable />
            </ExpandableButton>
          )}
        </ApplyBoxshadow>
      )}
      <DisplayExternalModal
        embed={embed}
        isEditMode={isEditMode}
        src={properties.src}
        type={properties.type}
        onEditEmbed={onEditEmbed}
        onClose={closeEditEmbed}
        allowedProvider={allowedProvider}
      />
      {children}
    </div>
  );
};

export default DisplayExternal;
