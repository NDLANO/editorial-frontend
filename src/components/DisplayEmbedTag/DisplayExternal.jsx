/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { injectT } from 'ndla-i18n';
import { Button } from 'ndla-ui';
import Types from 'slate-prop-types';
import { Cross, Pencil } from 'ndla-icons/action';
import './h5pResizer';
import Lightbox from '../Lightbox';
import VisualElementSearch from '../../containers/VisualElement/VisualElementSearch';
import handleError from '../../util/handleError';
import EditorErrorMessage from '../SlateEditor/EditorErrorMessage';
import { fetchExternalOembed } from '../../util/apiHelpers';
import { EditorShape } from '../../shapes';
import { editorClasses } from '../SlateEditor/plugins/embed/SlateFigure';
import { urlDomain } from '../../util/htmlHelpers';
import { EXTERNAL_WHITELIST_PROVIDERS } from '../../constants';

const el = document.createElement('html');

export const getIframeSrcFromHtmlString = html => {
  el.innerHTML = html;
  const iframe = el.getElementsByTagName('iframe')[0];
  return iframe.getAttribute('src');
};

export class DisplayExternal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editH5pMode: false,
    };
    this.onEditEmbed = this.onEditEmbed.bind(this);
    this.toggleEditH5p = this.toggleEditH5p.bind(this);
    this.getPropsFromEmbed = this.getPropsFromEmbed.bind(this);
  }

  componentDidMount() {
    this.getPropsFromEmbed(this.props.url);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.url !== this.props.url) {
      this.getPropsFromEmbed(this.props.url);
    }
  }

  onEditEmbed(properties) {
    const { editor, node } = this.props;

    if (properties.url !== this.props.url) {
      const next = editor.value.change().setNodeByKey(node.key, {
        data: {
          ...properties,
        },
      });
      editor.onChange(next);

      this.toggleEditH5p();
    }
  }

  async getPropsFromEmbed(url) {
    const { embed } = this.props;
    const domain = urlDomain(url);
    this.setState({ domain });

    if (embed.resource === 'external') {
      try {
        const data = await fetchExternalOembed(url);
        const src = getIframeSrcFromHtmlString(data.html);
        if (src) {
          this.setState({
            title: data.title,
            src,
            type: data.type,
            provider: data.providerName,
          });
        } else {
          this.setState({ error: true });
        }
      } catch (e) {
        handleError(e);
        this.setState({ error: true });
      }
    } else {
      this.setState({
        title: domain,
        src: url,
        type: embed.resource,
        height: embed.height,
      });
    }
  }

  toggleEditH5p() {
    const { changeVisualElement } = this.props;
    if (changeVisualElement) {
      changeVisualElement('h5p');
    } else
      this.setState(prevState => ({ editH5pMode: !prevState.editH5pMode }));
  }

  render() {
    const { onRemoveClick } = this.props;
    const { title, src, height, error, type, provider, domain } = this.state;

    if (!type && !provider) return null;

    // H5P does not provide its name
    const providerName = domain && domain.includes('h5p') ? 'H5P' : provider;

    const [allowedProvider] = EXTERNAL_WHITELIST_PROVIDERS.filter(
      whitelistProvider => {
        switch (type) {
          case 'iframe':
            return whitelistProvider.url.includes(domain);
          default:
            return whitelistProvider.name === providerName;
        }
      },
    );

    const renderProvider = allowedProvider && (
      <iframe
        ref={iframe => {
          this.iframe = iframe;
        }}
        src={src}
        height={allowedProvider.height || height}
        title={title}
        scrolling={type === 'iframe' ? 'no' : undefined}
        allowFullScreen={allowedProvider.fullscreen || true}
        frameBorder="0"
      />
    );

    if (error) {
      return (
        <Fragment>
          <Button
            stripped
            onClick={onRemoveClick}
            {...editorClasses('delete-button')}>
            <Cross />
          </Button>
          <EditorErrorMessage
            msg={this.props.t('displayOembed.errorMessage')}
          />
        </Fragment>
      );
    }

    return (
      <Fragment>
        <div {...editorClasses('figure-buttons', '')}>
          <Button
            stripped
            onClick={onRemoveClick}
            {...editorClasses('button', 'red')}>
            <Cross />
          </Button>
          {providerName === 'h5p' && (
            <Button
              stripped
              onClick={this.toggleEditH5p}
              {...editorClasses('button', 'green')}>
              <Pencil />
            </Button>
          )}
        </div>
        {renderProvider}
        {!renderProvider && (
          <EditorErrorMessage
            msg={this.props.t('displayOembed.notSupported', {
              type,
              provider,
            })}
          />
        )}
        {this.state.editH5pMode && (
          <Lightbox display fullscreen big onClose={this.toggleEditH5p}>
            <VisualElementSearch
              selectedResource="h5p"
              selectedResourceUrl={src}
              handleVisualElementChange={this.onEditEmbed}
              closeModal={this.toggleEditH5p}
            />
          </Lightbox>
        )}
      </Fragment>
    );
  }
}

DisplayExternal.propTypes = {
  url: PropTypes.string.isRequired,
  onRemoveClick: PropTypes.func,
  changeVisualElement: PropTypes.func,
  editor: EditorShape,
  node: Types.node,
  isIframe: PropTypes.bool,
  embed: PropTypes.shape({
    width: PropTypes.string,
    heigth: PropTypes.string,
    url: PropTypes.string,
  }),
};

export default injectT(DisplayExternal);
