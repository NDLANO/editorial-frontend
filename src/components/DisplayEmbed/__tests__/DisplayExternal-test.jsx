/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import React from 'react';
import { render } from '@testing-library/react';
import IntlWrapper from '../../../util/__tests__/IntlWrapper';
import DisplayExternal from '../DisplayExternal';
import { getIframeSrcFromHtmlString } from '../../../util/htmlHelpers';

function createNodeMock(element) {
  if (element.type === 'div') {
    return {
      querySelector: () => {},
    };
  }
  return null;
}
const options = { createNodeMock };

test('getIframeSrcFromHtmlString returns src attribute', () => {
  const src = getIframeSrcFromHtmlString(
    `<div>
      <iframe
        width="800"
        height="600"
        src="https://h5p-test.ndla.no/resource/3ab6850d">
      </iframe>
      <script src="https://h5p.org/h5p-resizer.js"></script>
    </div>`,
  );
  expect(src).toBe('https://h5p-test.ndla.no/resource/3ab6850d');
});

test('getIframeSrcFromHtmlString returns null id src not found', () => {
  const src = getIframeSrcFromHtmlString(
    `<div>
      <iframe
        width="800"
        height="600"
      </iframe>
      <script src="https://h5p.org/h5p-resizer.js"></script>
    </div>`,
  );
  expect(src).toBe(null);
});

test('DisplayExternal renders external correctly', () => {
  nock('http://ndla-api/')
    .persist()
    .get('/oembed-proxy/v1/oembed?url=https%3A%2F%2Fndla.no%2Foembed')
    .reply(200, {
      resource: 'external',
      title: 'unit test',
      providerName: 'Vimeo',
      type: 'rich',
      html: '<iframe src="iframe.html">',
    });

  const embed = {
    resource: 'external',
    url: 'https://ndla.no/oembed',
  };

  const  { container } = render(
    <IntlWrapper>
      <DisplayExternal embed={embed} />
    </IntlWrapper>,
    options,
  );
  expect(container).toMatchSnapshot();
});

test('DisplayExternal renders iframe correctly', () => {
  const embed = {
    height: '392',
    resource: 'iframe',
    url: 'https://nb.khanacademy.org/embed_video?v=jHPr-CuvHhs',
    width: '618',
  };

  const { container } = render(
    <IntlWrapper>
      <DisplayExternal onRemoveClick={() => ''} embed={embed} />
    </IntlWrapper>,
    options,
  );

  expect(container).toMatchSnapshot();
});

test('DisplayExternal display error on fetch fail', () => {
  nock('http://ndla-api/')
    .persist()
    .get('/oembed-proxy/v1/oembed?url=https%3A%2F%2Fndla.no%2Foembed')
    .replyWithError({
      message: 'something awful happened',
      code: 'AWFUL_ERROR',
    });

  const { container } = render(
    <IntlWrapper>
      <DisplayExternal embed={{ url: 'https://ndla.no/oembed' }} />
    </IntlWrapper>,
    options,
  );

  expect(container).toMatchSnapshot();
  setTimeout(() => {}, global.DEFAULT_TIMEOUT);
});
