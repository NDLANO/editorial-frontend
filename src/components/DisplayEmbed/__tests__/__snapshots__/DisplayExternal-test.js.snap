// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayExternal display error on fetch fail 1`] = `null`;

exports[`DisplayExternal renders external correctly 1`] = `null`;

exports[`DisplayExternal renders iframe correctly 1`] = `
.emotion-14 {
  position: absolute;
  top: 0;
  z-index: 1;
  right: -6.5px;
}

.emotion-14 > * {
  margin-bottom: 6.5px;
}

.emotion-5 {
  position: relative;
}

.emotion-1 {
  opacity: 0;
  position: absolute;
  z-index: 9999;
  pointer-events: none;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  -webkit-animation-delay: 0ms;
  animation-delay: 0ms;
  -webkit-animation-duration: 300ms;
  animation-duration: 300ms;
}

.emotion-0 {
  display: block;
  color: #fff;
  background: #20588f;
  border-radius: 2px;
  padding: 6.5px 13px;
  font-family: 'Source Sans Pro',Helvetica,Arial,STKaiti,'华文楷体',KaiTi,SimKai,'楷体',KaiU,DFKai-SB,'標楷體',SongTi,'宋体',sans-serif;
  font-size: 14;
  font-size: 0.7777777777777778rem;
  line-height: 1.2 font-weight:400;
  color: $white;
  text-align: center;
  white-space: nowrap;
  max-width: calc(100vw - #{26px});
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0;
  padding: 0;
  background: 0;
  border: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #d1372e99;
  width: 26px;
  height: 26px;
}

.emotion-3 svg {
  width: 22px;
  height: 22px;
  position: relative;
  z-index: 1;
  pointer-events: none;
}

.emotion-3::before {
  content: '';
  display: block;
  position: absolute;
  width: 26px;
  height: 26px;
  background: #d1372e4D;
  border-radius: 100%;
  -webkit-transform: scale(0.5);
  -ms-transform: scale(0.5);
  transform: scale(0.5);
  opacity: 0;
  -webkit-transition: all 200ms ease;
  transition: all 200ms ease;
}

.emotion-3:hover,
.emotion-3:focus {
  color: #d1372e;
}

.emotion-3:hover::before,
.emotion-3:focus::before {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  opacity: 1;
}

.emotion-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0;
  padding: 0;
  background: 0;
  border: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #a5bcd3;
  width: 26px;
  height: 26px;
}

.emotion-10 svg {
  width: 22px;
  height: 22px;
  position: relative;
  z-index: 1;
  pointer-events: none;
}

.emotion-10::before {
  content: '';
  display: block;
  position: absolute;
  width: 26px;
  height: 26px;
  background: #deebf6;
  border-radius: 100%;
  -webkit-transform: scale(0.5);
  -ms-transform: scale(0.5);
  transform: scale(0.5);
  opacity: 0;
  -webkit-transition: all 200ms ease;
  transition: all 200ms ease;
}

.emotion-10:hover,
.emotion-10:focus {
  color: #20588f;
}

.emotion-10:hover::before,
.emotion-10:focus::before {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  opacity: 1;
}

<div
  className="c-figure"
>
  <div
    className="emotion-14 emotion-15"
  >
    <div
      className=" emotion-5 emotion-6"
    >
      <div
        className="emotion-1 emotion-2"
      >
        <span
          className="emotion-0"
          role="tooltip"
          style={Object {}}
        >
          
        </span>
      </div>
      <div
        aria-label=""
        className=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="emotion-3 emotion-4"
          color="red"
          data-cy="remove-element"
          onClick={[Function]}
          tabIndex={-1}
          type="button"
        >
          <svg
            className="c-icon"
            data-license="Apache License 2.0"
            data-source="Material Design"
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "color": undefined,
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 24 24"
            width="1em"
          >
            <title>
              DeleteForever
            </title>
            <g>
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </g>
          </svg>
        </button>
      </div>
    </div>
    <div
      className=" emotion-5 emotion-6"
    >
      <div
        className="emotion-1 emotion-2"
      >
        <span
          className="emotion-0"
          role="tooltip"
          style={Object {}}
        >
          
        </span>
      </div>
      <div
        aria-label=""
        className=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="emotion-10 emotion-4"
          onClick={[Function]}
          tabIndex={-1}
        >
          <svg
            className="c-icon"
            data-license="Apache License 2.0"
            data-source="Material Design"
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style={
              Object {
                "color": undefined,
                "verticalAlign": "middle",
              }
            }
            viewBox="0 0 24 24"
            width="1em"
          >
            <title>
              Link
            </title>
            <g>
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
              />
            </g>
          </svg>
        </button>
      </div>
    </div>
  </div>
  <iframe
    allowFullScreen={true}
    frameBorder="0"
    height="486px"
    scrolling="no"
    src="https://nb.khanacademy.org/embed_video?v=jHPr-CuvHhs"
    title="nb.khanacademy.org"
  />
</div>
`;
